src/query_plan/query_planning_traversal.rs

```diff
@@ -86,3 +88,116 @@ struct BestQueryPlanInfo {
     /// The cost of this query plan.
     cost: QueryPlanCost,
 }
+
+impl QueryPlanningTraversal {
+    pub(crate) fn new_dependency_graph(&self) -> FetchDependencyGraph {
+        let root_type = if self.is_top_level && self.has_defers {
+            self.parameters
+                .supergraph_schema
+                .schema()
+                .root_operation(self.root_kind.into())
+                .cloned()
+                // A root operation type has to be an object type
+                .map(|type_name| ObjectTypeDefinitionPosition { type_name }.into())
+        } else {
+            None
+        };
+        FetchDependencyGraph::new(
+            self.parameters.supergraph_schema.clone(),
+            self.parameters.federated_query_graph.clone(),
+            root_type,
+            self.starting_id_generation,
+        )
+    }
+
+    fn updated_dependency_graph(
+        &self,
+        dependency_graph: &mut FetchDependencyGraph,
+        path_tree: &OpPathTree,
+    ) -> Result<(), FederationError> {
+        // TODO: is this right?
+        // In `query-planner-js/src/buildPlan.ts`, `isRootPathTree(tree)` is dependent on `tree`,
+        // but type-Tetris gives us something dependent on `self.root_kind`.
+        let is_root_path_tree = self
+            .parameters
+            .federated_query_graph
+            .root_kinds_to_nodes()?
+            .contains_key(&self.root_kind);
+
+        if is_root_path_tree {
+            // The root of the pathTree is one of the "fake" root of the subgraphs graph,
+            // which belongs to no subgraph but points to each ones.
+            // So we "unpack" the first level of the tree to find out our top level groups
+            // (and initialize our stack).
+            // Note that we can safely ignore the triggers of that first level
+            // as it will all be free transition, and we know we cannot have conditions.
+            for child in &path_tree.childs {
+                let edge = child.edge.expect("The root edge should not be None");
+                let (_source_node, target_node) = path_tree.graph.edge_endpoints(edge)?;
+                let target_node = path_tree.graph.node_weight(target_node)?;
+                let subgraph_name = &target_node.source;
+                // The edge tail type is one of the subgraph root type, so it has to be an ObjectType.
+                //     const rootType = edge.tail.type as ObjectType;
+                let root_type = match &target_node.type_ {
+                    QueryGraphNodeType::SchemaType(position) => match position {
+                        OutputTypeDefinitionPosition::Object(object) => object.clone().into(),
+                        _ => todo!(), // Return a FederationError?
+                    },
+                    QueryGraphNodeType::FederatedRootType(_) => todo!(), // Return a FederationError?
+                };
```

Comment:
TypeScript `getOrCreateRootFetchGroup` expects `CompositeType` which I assume corresponds to Rust enum `CompositeTypeDefinitionPosition`.

This code obtains `QueryGraphNodeType`. How should the `FederatedRootType` case be handled?

Reply:
The `FederatedRootType`-case only occurs for root nodes of the federated query graph, so you can return error here in that case (children of federated root nodes should be `SchemaType` nodes, i.e. they should belong to some subgraph schema).